File: DataStructure/StackQueue/SqStack.cpp
/*
    作者: btx
    日期: 23-5-30
    功能：顺序栈的基本实现
*/
#include <stdio.h>
#include <stdlib.h>
#define ElemType int
#define MaxSize 10
typedef struct{
    ElemType data[MaxSize];
    int top; //栈顶指针
}SqStack;
//初始化
void srackInit(SqStack &S){
    S.top = -1;
}
//入栈
bool push(SqStack &S,ElemType x){
    //判断栈是否已满
    if(S.top == MaxSize-1)
        return false;
    S.top = S.top + 1;//指针+1
    S.data[S.top] = x;//入栈
    return true;
}
bool pop(SqStack &S,ElemType &x){
    if(S.top == -1)
        return false;
    x = S.data[S.top];//先出栈
    S.top = S.top - 1;//指针减1
    return true;
}
//获取栈顶元素
bool getTop(SqStack &S,ElemType &x){
    if(S.top == -1)
        return false;
    x = S.data[S.top];
    return true;
}
int main(){
    SqStack S;
    srackInit(S);
    push(S,1);
    push(S,2);
    push(S,3);
    int x;
    getTop(S,x);
    printf("当前栈顶元素为:%d\n",x);
    pop(S,x);
    printf("出栈:%d\n",x);
    pop(S,x);
    printf("出栈:%d\n",x);
    pop(S,x);
    printf("出栈:%d\n",x);
    return 0;
}

File: DataStructure/StackQueue/SqStack.py
'''
    作者: btx
    日期: 23-5-31
    功能：顺序栈的基本实现
'''
class Stack:
    def __init__(self):
        self.stack = []
    #入栈
    def push(self, item):
        self.stack.append(item)
    #出栈
    def pop(self):
        if not self.is_empty():
            return self.stack.pop()
    #获取栈顶元素
    def getTop(self):
        if not self.is_empty():
            return self.stack[-1]

    def is_empty(self):
        return len(self.stack) == 0

    def size(self):
        return len(self.stack)
    
if __name__ == "__main__":
    stack = Stack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    print("输出当前栈的大小")
    print(stack.size())
    print("获取栈顶元素")
    print(stack.getTop())
    print("依次出栈")
    print(stack.pop())
    print(stack.pop())
    print(stack.pop())
    print(stack.pop())
